### DISCUSSION TITLE
Increase in inference speed with 2 TPUs vs 1 TPU

### DISCUSSION THREAD

**Discussion by Cold-Lemonade** [blakeblackshear/frigate#4125] [General]
I am running Frigate 0.11.1 in a docker container on a dedicated Minisforum GK41, with 8 GB of memory, running Ubuntu Server 22.04. I have 11 cameras (7 ip cams and 4 wyze cams). My inference speed was as high as 13-14 fps at times. So I recently added a second Google Coral usb. My inference speed hasn't changed by much. The Frigate debug page shows each of the two usb has an inference speed of about 10-12 fps. Does that sound right? Can I do anything else speed things up?

**Comment by NickM-27** [comment]
The inference speeds is not in fps it is in milliseconds. 

10-12 milliseconds is expected for the coral and is quite fast

**Reply by banthungprong** [reply]
I have exactly the same HW/SW configuration: GK41, Ubuntu 22.04 and 2 TPUs. 8 Cameras.
Just checked my values: The values of the inference speed aren't constant. Range between 11 and 17.

**Reply by banthungprong** [reply]
And now it is getting funny:
I removed one Coral, now the values are between 10-11. Will test again with 2 devices and compare. 
So one TPU for my cameras seems to be enough. Adding one more might not help at all - maybe limit with the USB ports?!?
More test didn't reproduce the values above. Alway range between 10-11. 
In the end: with my 8 cameras and detection fps 5 to 10 I do need only one TPU

**Reply by banthungprong** [reply]
From the Coral Website:

> Before you add more Edge TPUs in your system, consider the following possible performance issues:

> Python does not support real multi-threading for CPU-bounded operations (read about the [Python global interpreter lock (GIL)](https://docs.python.org/3.5/c-api/init.html#thread-state-and-the-global-interpreter-lock)). However, we have optimized the Edge TPU Python API (but not TensorFlow Lite Python API) to work within Python’s multi-threading environment for all Edge TPU operations—they are IO-bounded, which can provide performance improvements. But beware that CPU-bounded operations such as image downscaling will probably encounter a performance impact when you run multiple models because these operations cannot be multi-threaded in Python.

> When using multiple USB Accelerators, your inference speed will eventually be bottlenecked by the host USB bus’s speed, especially when running large models.

So I stay with one TPU and use the second for development or a later second device in Thailand...

**Reply by blakeblackshear** [reply]
> From the Coral Website:
> 
> > Before you add more Edge TPUs in your system, consider the following possible performance issues:
> 
> > Python does not support real multi-threading for CPU-bounded operations (read about the [Python global interpreter lock (GIL)](https://docs.python.org/3.5/c-api/init.html#thread-state-and-the-global-interpreter-lock)). However, we have optimized the Edge TPU Python API (but not TensorFlow Lite Python API) to work within Python’s multi-threading environment for all Edge TPU operations—they are IO-bounded, which can provide performance improvements. But beware that CPU-bounded operations such as image downscaling will probably encounter a performance impact when you run multiple models because these operations cannot be multi-threaded in Python.
> 
> > When using multiple USB Accelerators, your inference speed will eventually be bottlenecked by the host USB bus’s speed, especially when running large models.
> 
> So I stay with one TPU and use the second for development or a later second device in Thailand...

The GIL is not a limitation for the way Frigate is architected. Each Coral runs in a dedicated separate process and they share a multiprocessing queue. It is possible that your USB bus could be maxed out. On my development laptop, I can hit double the detection fps with two Corals. I would recommend using both as it will help when multiple cameras need detection simultaneously.

**Reply by banthungprong** [reply]
That was the idea. But right now 1 TPU is enough, so I can use the other one connected to my PC when playing with frigate sources in VS Code :smiley: (adding clip length in Event View and option for enabling/disabling a cam in the yaml was already fun).

I raised the detection fps on all 8 cameras to 10 (one with 15) and still everything seems to be ok.

**Comment by Cold-Lemonade** [comment]
Yes. I meant milliseconds--I don't know what I was thinking other than it must have been pretty late for me.

@banthungprong Are you using the USB 3.0 ports? I have not seen my inference speed rise to 17 milliseconds. It fluctuates between 10 and 11.

**Reply by banthungprong** [reply]
Only in the first check. After switching to other ports and back again the values were in the normal range. Will check in the next few days if these values raise after some time.

**Comment by AML225** [comment]
I have 2x dual edge TPUs connected via PCIe with magic-blue-smoke's adapter and get inference speeds in the 8-10ms range. My previous setup used a single USB TPU and was in the 10-15ms range. 

![image](https://user-images.githubusercontent.com/37520295/197315749-30493660-38a1-49b7-bf41-99e39bf797ed.jpeg)

**Reply by blakeblackshear** [reply]
That's a dream setup.

**Reply by AML225** [reply]
> That's a dream setup. 

The man himself! Thank you for your hard work!

Honestly it just took patience, really wasn't terribly expensive:

- Used Dell 7060 w/ i7-8700 & 8gb RAM: $300
- 2x Dual Edge TPU adapters = $60
- 2x Dual Edge TPUs = $80
Total Cost: $440

Took about 12 months of waiting for the TPUs though.

**Reply by NickM-27** [reply]
I've got one dual and one single, been working great and obviously a lot of headroom. 

![Screen Shot 2022-10-24 at 16 59 18 PM](https://user-images.githubusercontent.com/14866235/197645709-dc457b01-92c5-4efd-9e60-e6d56c36546e.png)

**Reply by FrigateEnthusiast** [reply]
Those temps look great but just want to toss this out as a reminder that the TPU will do throttling once temp trip points are hit. First one is 85C.

"The Edge TPU runtime. (Pre-installed on the Dev Board.) The separate versions affect only USB-based devices to determine the operating frequency: libedgetpu1-max runs at the maximum frequency (500 MHz) and libedgetpu1-std runs at the reduced frequency (250 MHz).
For PCIe-based devices, both packages behave the same because the PCIe driver performs [dynamic frequency scaling](https://coral.ai/docs/pcie-parameters/#use-dynamic-frequency-scaling) to adjust the operating frequency based on programmable temperature thresholds."

https://coral.ai/docs/pcie-parameters/#use-dynamic-frequency-scaling

I just ordered small copper heatsinks with tall fins to put on my EdgeTpu's so I never have to worry about hitting throttle points.

**Reply by bracklanna** [reply]
At the risk of necro'ing a thread -- what copper heatsinks did you order? Do you have a link to them or to something similar?  If you followed through on that, how did they / are they working out for you?

I'm running a single Coral USB TPU on a Dell Precision 3420 tower (6th-gen i5-6600 3.2Ghz, 8GB RAM) seeing inference speeds around 10ms. Very tasty. No issues with excessive heat so far but I'd like to future-proof in case of hot weather & lots of motion in front of cameras pushing up the Coral TPU load.

**Reply by FrigateEnthusiast** [reply]
These aren't the exact ones but just an idea of what's out there (lots of them). I always like to buy extra and have them ready to go to as needed... As far as how "did they work out...", they simply work.

https://www.amazon.com/ZDE-Heatsinks-Conductivity-Dissipation-Compatible/dp/B0DKBP3NH7/ref=sr_1_4

**Comment by Jimmy-SafeCash** [comment]
I have a dual tpu passed through to a kubernetes node via proxmox. I have a constant 8-9 inference speed. Very impressed with what an old dell r730 can do.

**Reply by shawnventi** [reply]
I'm also running on an older Dell R730 and was just able to install my single Coral TPU today. My inference speed has been just like yours around 8-9 ms all morning. I have 20 Reolink cameras on this deployment.

How many and what type of camera feeds do you have on your setup?

**Reply by FrigateEnthusiast** [reply]
Is that 8-9 for usb or pcie?

**Reply by tho1981** [reply]
I suppose the PCIE can. be expected to deliver a but better than the USB version?

**Comment by drocanyd** [comment]
I'm using a single PCIe coral with 16 Dahua cameras and getting inference averaging 8-9 consistently with detection at 3fps.

machine info:

CPU(s):                          40
On-line CPU(s) list:             0-39
Thread(s) per core:              2
Core(s) per socket:              10
MemTotal:       82382088 kB
